Add classes for collision checking

bpwilcox opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->
This PR adds classes to abstract the footprint and costmap2D across the ROS2 topic interface and enable collision checking in other modules. The collision checking code is modified from the `ObstacleFootprintCritic` methods

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #533  |

---

## Description of contribution in a few bullet points
- Publishes a `nav2_msgs::msg::Costmap` in the `CostmapPublisher`
- Adds a `FootprintSubscriber` class
- Adds a `CostmapSubscriber` class
- Adds a `CollisionChecker` class
- Adds a test for the `CollisionChecker`
- duplicates `line_iterator.hpp` in `nav2_util` (to be later removed from `dwb_critics`)


## Future work
- Use above classes in recovery behaviors
- Refactor dwb_controller to utilize subscriber classes and collision checker

SteveMacenski commented on nav2_costmap_2d/src/footprint_subscriber.cpp:
```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2019 Intel Corporation
```
I see no issue with this, but I would believe that having this read the configuration file and it being a ROS agnostic helper to read from configuration file and getting as necessary may be more clear since somewhere else inside of costmap_2d we read the footprint and then publish it. Seems odd to have things running in the same process in the same node talking to each other over ROS to share data

bpwilcox replied:
I believe in `Costmap2DROS`, the footprint subscription is listening on a Polygon topic in order to update the footprint across all layers via `setUnpaddedFootprint` in its callback. Since in this class we're listening on the published PolygonStamped topic and have a default callback to convert and store the robot oriented footprint, I see them as separate use cases.

bpwilcox commented on nav2_costmap_2d/src/collision_checker.cpp:
```diff
@@ -0,0 +1,209 @@
+
+CollisionChecker::~CollisionChecker() {}
+
+bool CollisionChecker::isCollisionFree(
```
I'll look and give some thought into returning enum types. The exceptions here are modified from those in DWB as you mentioned.

mjeronimo replied:
IMO exceptions should be used when the local code can't recover from the error but someone up the stack may be able to. They shouldn't be used to propagate a normal return value. For example, if isCollisionFree may commonly return true or false, a return value is more appropriate. So, both exceptions and return values are appropriate, depending on the circumstances, and should be applied the right way.

bpwilcox replied:
I could modify `scorePose` to return a boolean and replace  `isCollisionFree`, but in this PR I was moving the code (including the exceptions) from the `ObstacleFootprint` critic which returns a score for the pose requested. My intention was to replace the `ObstacleFootprint` critic code with this class to avoid code duplication, which is why I thought to keep it as is.

bpwilcox commented on nav2_costmap_2d/src/collision_checker.cpp:
```diff
@@ -0,0 +1,209 @@
+  double footprint_cost = 0.0;
+
+  // we need to rasterize each line in the footprint
+  for (unsigned int i = 0; i < footprint.size() - 1; ++i) {
```
That would be a more robust check I agree, though for this PR it matches equivalence with ROS1's collision checking during MoveBase recoveries (uses same algorithm from obstacle footprint critic). I think a future PR can add BFS or something.

SteveMacenski replied:
Ok. Added a ticket here: https://github.com/ros-planning/navigation2/issues/740

bpwilcox commented on nav2_costmap_2d/src/collision_checker.cpp:
```diff
@@ -0,0 +1,209 @@
+}
+
+bool
+CollisionChecker::getRobotPose(geometry_msgs::msg::PoseStamped & global_pose) const
```
I agree. We should move the one in `Costmap2dROS` as well.

orduno commented on nav2_costmap_2d/src/collision_checker.cpp:
```diff
@@ -0,0 +1,209 @@
+  // we need to rasterize each line in the footprint
+  for (unsigned int i = 0; i < footprint.size() - 1; ++i) {
+    // get the cell coord of the first point
+    if (!costmap_->worldToMap(footprint[i].x, footprint[i].y, x0, y0)) {
```
Making these checks a separate function might reduce some duplication `worldToMap(wx, wy, mx, mw)`

orduno replied:
and maybe have this iteration over all points inside an other function `validateFootprint`.

bpwilcox replied:
This for loop does more than just validate the footprint, it also gets the map coordinates which are passed through to `lineCost(x0, x1, y0, y1)` for each point pair.

orduno commented on nav2_costmap_2d/src/collision_checker.cpp:
```diff
@@ -0,0 +1,209 @@
+
+double CollisionChecker::pointCost(int x, int y)
+{
+  Costmap2D * costmap_ = costmap_sub_->getCostmap();
```
`costmap_` -> `costmap`

orduno commented on nav2_costmap_2d/src/costmap_subscriber.cpp:
```diff
@@ -0,0 +1,81 @@
+void CostmapSubscriber::toCostmap2D()
+{
+  if (!costmap_received_) {
+    costmap_ = new Costmap2D(
```
Where to we `delete`? alternatively use `std::shared_ptr`

mjeronimo replied:
Or unique_ptr if it is not shared.

orduno commented on nav2_costmap_2d/src/collision_checker.cpp:
```diff
@@ -0,0 +1,209 @@
+double CollisionChecker::scorePose(
+  const geometry_msgs::msg::Pose2D & pose)
+{
+  Costmap2D * costmap_;
```
Why have a costmap pointer? Maybe move the `worldToMap` checks to `CostmapSubscriber`

bpwilcox replied:
Because the intention of the `CostmapSubscriber` was to convert the msg into the `Costmap2D` object to access its API calls. I think it's better not to duplicate that code in the subscriber class

orduno commented on nav2_costmap_2d/src/costmap_subscriber.cpp:
```diff
@@ -0,0 +1,81 @@
+      std::bind(&CostmapSubscriber::costmap_callback, this, std::placeholders::_1));
+}
+
+Costmap2D * CostmapSubscriber::getCostmap()
```
Instead of providing the costmap to the clients, can we define an interface that provides what the clients really need, i.e. poseWithinMap(), poseCost(), etc.

orduno replied:
I think the costmap_subscriber would eventually become the client:

![client](https://user-images.githubusercontent.com/39749557/52814434-04a64300-3051-11e9-93e1-1acbcc4e7840.png)

bpwilcox replied:
Perhaps, though I would address this as a future feature of the world model in a different PR. Also, I'm not sure how much client-specific context we'd really like at the interface level. Calling something like `poseCost()` is very context-specific and I'm not sure in the scope of the WorldModel. In my view, the WorldModel could be a plugin container for abstract map representations (of which costmap2D is one) of the environment from which clients can grab and use as needed on their end. We can continue this discussion

orduno commented on nav2_costmap_2d/include/nav2_costmap_2d/footprint_subscriber.hpp:
```diff
@@ -0,0 +1,48 @@
+
+}  // namespace nav2_costmap_2d
+
+#endif  // NAV2_COSTMAP_2D__FOOTPRINT_SUBSCRIBER_HPP_
```
No newline at end of file, also so this on other files.

orduno commented on nav2_costmap_2d/include/nav2_costmap_2d/exceptions.hpp:
```diff
@@ -0,0 +1,69 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2017, Locus Robotics
```
Add:
Copyright (c) 2019, Intel Corporation

crdelsey replied:
This looks like mostly copied code. I wouldn't change the copyright.

orduno commented:
Mostly style recommendations, just one or two things that need to be fixed before merging.

bpwilcox commented on nav2_costmap_2d/src/collision_checker.cpp:
```diff
@@ -0,0 +1,209 @@
+  const geometry_msgs::msg::Pose2D & pose)
+{
+  try {
+    if (scorePose(pose) < 0) {
```
Yes, this is according to the code in `ObstacleFootprintCritic`

bpwilcox commented on nav2_costmap_2d/src/collision_checker.cpp:
```diff
@@ -0,0 +1,209 @@
+    throw CollisionCheckerException(e.what());
+  }
+
+  unsigned int cell_x, cell_y;
```
To get the bool flag, no we don't need the cell_x, cell_y, though I think it is useful for debugging and it is part of the `Costmap2D` API. I'm opposed to doing the check inside `CostmapSubscriber` because that class should be general for anyone subscribing to the costmap topic, not just for collision checking

bpwilcox commented:
I've addressed feedback, so this PR is open again for review. I'm adding the WIP to note that I am waiting on PR #793 to be merged so I can get rid of the tf_buffer to grab the robot pose and instead use the service call. Right now, I've hard-coded the global frames so that I don't have to pass in a node in the `CollisionChecker` just to grab params off the node.

bpwilcox commented:
Open again for review now that #793 is merged and modifications were made.

mjeronimo commented on nav2_costmap_2d/include/nav2_costmap_2d/collision_checker.hpp:
```diff
@@ -0,0 +1,66 @@
+#include "nav2_costmap_2d/footprint_subscriber.hpp"
+#include "nav2_util/get_robot_pose_client.hpp"
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpedantic"
```
Not gating this PR, but perhaps you can submit an issue to fix this upstream so we don't have to work around the warnings.

mjeronimo commented on nav2_costmap_2d/include/nav2_costmap_2d/collision_checker.hpp:
```diff
@@ -0,0 +1,66 @@
+
+  double scorePose(
+    const geometry_msgs::msg::Pose2D & pose);
+  bool isCollisionFree(
```
Why the line breaks for these methods? Makes for inconsistent formatting.

orduno commented:
@bpwilcox is this PR replaced by #889?

bpwilcox commented:
@orduno This PR was intended to be a prerequisite for #889 (stated in its description) as it is built off of this PR. Since there was feedback on the same commits from this PR, I'm just addressing all the feedback here.

bpwilcox commented:
Closing because #889 is merged.
